### PR TITLE
Usecases 1ii 1iii

### PR DESCRIPTION
**Pull request by hsorby**
addresses issue #25

### CODE REVIEW

**Comment by nickerso** [comment]
The tests look good to me.

**Comment by nickerso** [comment]
@agarny @jonc125 @mirams I have added a test for "unsetting" the name showing how I expect things to behave (which is not how @hsorby has initially implemented it). What do you think? I imagine similar behaviour for unsetting cmeta:id's once they are implemented...

File: src/model.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@ limitations under the License.Some license of other
 
 namespace libcellml {
 
+static const std::string initial_name = "OhOXiRmNswYoVBA2InM7NRPvQ0HMXFR2wNhoFExd";
```
Comment (agarny) [inline src/model.cpp]: Why this? Why if someone was to use exactly that name for a model? Why not simply have an empty string?

File: src/model.cpp [hunk 2]
```diff
@@ -60,11 +64,25 @@ std::string Model::serialise(libcellml::CELLML_FORMATS format) const
     std::string repr = "";
     if (format == CELLML_FORMAT_XML)
     {
-        repr += "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\"></model>";
+        repr += "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\"";
+        if (mPimpl->mName != initial_name)
+        {
```
Comment (agarny) [inline src/model.cpp]: We want `{` at the end of the previous line (see https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Conditionals).

File: tests/model/model.cpp [hunk 3]
```diff
@@ -2,12 +2,50 @@
...
+    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"name\"></model>";
+
+    libcellml::Model m = libcellml::Model();
+    std::string n = "name";
+    m.setName(n);
+
+    EXPECT_EQ("name", m.getName());
+
+    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
+}
```
Comment (agarny) [inline tests/model/model.cpp:24]: I would rather have:

```
TEST(Model, name) {
    const std::string n = "name";
    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\""+n+"\"></model>";

    libcellml::Model m = libcellml::Model();
    m.setName(n);

    EXPECT_EQ(n, m.getName());
    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
}
```

File: tests/model/model.cpp [hunk 4]
```diff
@@ -2,12 +2,50 @@
...
+    libcellml::Model m = libcellml::Model();
+    std::string n = "name";
+    m.setName(n);
+    EXPECT_EQ("name", m.getName());
+    EXPECT_EQ(eName, m.serialise(libcellml::CELLML_FORMAT_XML));
+
+    m.setName("");
+    EXPECT_EQ("", m.getName());
+    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
+}
```
Comment (agarny) [inline tests/model/model.cpp:39]: Similarly, I would have something like:

```
TEST(Model, unset_name) {
    const std::string n = "name";
    const std::string eName = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\""+n+"\"></model>";
    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\"></model>";

    libcellml::Model m = libcellml::Model();
    m.setName(n);
    EXPECT_EQ(n, m.getName());
    EXPECT_EQ(eName, m.serialise(libcellml::CELLML_FORMAT_XML));

    m.setName("");
    EXPECT_EQ("", m.getName());
    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
}
```

File: tests/model/model.cpp [hunk 5]
```diff
@@ -2,12 +2,50 @@
...
+    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"invalid name\"></model>";
+
+    libcellml::Model m = libcellml::Model();
+    std::string in = "invalid name";
+    m.setName(in);
+
+    EXPECT_EQ("invalid name", m.getName());
+
+    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
+}
```
Comment (agarny) [inline tests/model/model.cpp:51]: And finally:

```
TEST(Model, invalid_name) {
    const std::string in = "invalid name";
    const std::string e = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\""+in+"\"></model>";

    libcellml::Model m = libcellml::Model();
    m.setName(in);

    EXPECT_EQ(in, m.getName());
    EXPECT_EQ(e, m.serialise(libcellml::CELLML_FORMAT_XML));
}
```

**Comment by agarny** [comment]
@nickerso, your test for "unsetting" the name is also how I would do it (as opposed to using `initial_name`, with `initial_name` being equal to `OhOXiRmNswYoVBA2InM7NRPvQ0HMXFR2wNhoFExd`, if that's what you meant).

**Comment by nickerso** [comment]
I'm happy with these changes now. I have no real preference for testing the name value and would be happy to leave it as it is or switch to Alan's preferred style.

**Pull request merged**